allensdk/brain_observatory/vbn_2022/metadata_writer/dataframe_manipulations.py

```diff
@@ -0,0 +1,422 @@
+
+            new_data.append(
+                {'behavior_session_id': beh_id,
+                 'prior_exposures_to_omissions': n_omissions})
+
+    new_data = pd.DataFrame(data=new_data)
+    behavior_df = behavior_df.join(
+            new_data.set_index('behavior_session_id'),
+            on='behavior_session_id',
+            how='left')
+    return behavior_df
+
+
+def _add_experience_level(
+        sessions_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Add the column 'experience_level' to a dataframe. This column
+    will be 'Novel' for any rows with
+    'prior_exposures_to_image_set' == 0 (or NULL) and 'Familiar'
+    otherwise.
+
+    Return the same dataframe with the column added
+    """
+
+    sessions_df['experience_level'] = np.where(
+                  np.logical_or(
+                      sessions_df['prior_exposures_to_image_set'] == 0,
+                      sessions_df['prior_exposures_to_image_set'].isnull()),
+                  'Novel',
+                  'Familiar')
+    return sessions_df
+
+
+def _patch_date_and_stage_from_pickle_file(
+        lims_connection: PostgresQueryMixin,
+        behavior_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Fill in missing date_of_acquisition and session_type
+    directly from the stimulus pickle file
+
+    Parameters
+    ----------
+    lims_connection: PostgresQueryMixin
+
+    behavior_df: pd.DataFrame
+        The dataframe to be patched
+
+    Returns
+    -------
+    behavior_df: pd.DataFrame
+        Identical to behavior_df, except any rows with NULL
+        date_of_acquisition or foraging_id will have their
+        date_of_acquisition and session_type overwritten with
+        values from the stimulus pickle file.
+    """
+
+    invalid_beh = behavior_df[
+            np.logical_or(
+                behavior_df.date_of_acquisition.isna(),
+                np.logical_or(
+                    behavior_df.foraging_id.isna(),
+                    behavior_df.session_type.isna()))
+    ].behavior_session_id.values
+
+    assert len(invalid_beh) == len(np.unique(invalid_beh))
+
+    if len(invalid_beh) > 0:
+        pickle_path_df = stimulus_pickle_paths_from_behavior_session_ids(
+                            lims_connection=lims_connection,
+                            behavior_session_id_list=invalid_beh)
+
+        for beh_id, pkl_path in zip(pickle_path_df.behavior_session_id,
+                                    pickle_path_df.pkl_path):
+            stim_file = BehaviorStimulusFile(filepath=pkl_path)
+            new_date = stim_file.date_of_acquisition
+            new_session_type = stim_file.session_type
+            behavior_df.loc[
+                behavior_df.behavior_session_id == beh_id,
+                ('date_of_acquisition', 'session_type')] = (new_date,
+                                                            new_session_type)
+
+    return behavior_df
+
+
+def _add_age_in_days(df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
hm... I didn't realize there as an `age` code in `donors`. I don't know how there could be. I'm pretty sure age_in_days is meant to be "age of the mouse at this session." Since each mouse is subjected to multiple sessions, this column should be different for different mice at different sessions. I wonder if the age column in donors is meant for surgically removed samples...